torch/csrc/jit/python/python_tracer.cpp

```diff
@@ -73,6 +73,58 @@ SourceRange getPythonInterpreterSourceRange() {
   return SourceRange(source, 0, stack_trace_text.size());
 }
 
+std::pair<std::shared_ptr<Graph>, Stack> createGraphByTracingWithDict(
+    const py::function& func,
+    const py::dict& inputs_dict,
+    Stack trace_inputs,
+    const py::function& var_name_lookup_fn,
+    bool strict,
+    bool force_outplace,
+    Module* self,
+    const std::vector<std::string>& argument_names) {
+  C10_LOG_API_USAGE_ONCE("torch.tracer");
+
+  auto lookup_fn_adapter =
+      [var_name_lookup_fn](const Variable& var) -> std::string {
+    pybind11::gil_scoped_acquire ag;
+    return py::cast<std::string>(var_name_lookup_fn(var));
+  };
+
+  // The argument_names parameter is parsed in python and its order
+  // is the same as the arguments' decalaration order in forward() method.
+  // These name shall be added to the graph as debug name and the order
+  // should align with the traceable stack we generated by the python dict.
+  std::vector<std::string> reordered_argument_names;
+  for (auto it = inputs_dict.begin(); it != inputs_dict.end(); it++) {
+    for (size_t i = 0; i < argument_names.size(); i++) {
+      if (py::cast<std::string>(it->first) == argument_names[i]) {
+        reordered_argument_names.push_back(argument_names[i]);
+        break;
+      }
+    }
+  }
```

Comment:
Could we reorder the inputs instead of the argument names? This might make more sense to people when they trace a model, so that their arguments don't get reordered compared to the method definition.

Reply:
It seems feasible. I will try it locally, and if I verify the result, I will follow your suggestion.

Reply:
Hi, @davidberard98 
I think we have to do this reordering. 
According to this [code snippet](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/tracer.cpp#L490-L508):
```
// When enough argument name hints are provided, use them as debug names
// for traced function/modules.
// Here argument_names is allowed to have more names than needed because
// some arguments may have valid default values, therefore they don't need
// example inputs.
if (argument_names.size() >= inputs.size()) {
  for (size_t i = 0, e = inputs.size(); i < e; ++i) {
    IValue& input = inputs[i];
    input = addInput(
        state,
        input,
        input.type(),
        state->graph->addInput(argument_names[i]));
  }
} else {
  for (IValue& input : inputs) {
    input = addInput(state, input, input.type(), state->graph->addInput());
  }
}
```
Pytorch will set the debug name of the traced torchscript according to the real input's order, in fact, I think what matters is the number of the real input other than the order. Use the example I used in the UT and RFC's Problem 2 in the first section, there was missing arguments, which cause  
the input and argments_name not match with each other. So, one of the intention here is to compact the argument_name and make them align with inputs in length.(I think it's impossible(at least I haven't come out a decent method) to expand the input with the default value of the missing args.), and another is reorder.
The debug name will be used in [_check_trace()](https://github.com/pytorch/pytorch/blob/master/torch/jit/_trace.py#L310) after the tracing procedure.

Reply:
Hmm, I think I understand... Can you do something like this instead?

```c++
for (i in 0..argument_names.size()) {
  if (argument_names[i] in input_dict) {
    reordered_argument_names.push_back(argument_names[i])
  }
}
```

Or am I misunderstanding the problem?

Reply:
[Commented below](https://github.com/pytorch/pytorch/pull/81623#issuecomment-1269542950)

Reply:
Hi, @davidberard98 
Do you mean put this logic in python other than C++?

Reply:
Hi, @davidberard98 
This way will not reorder the arguments, just compact. I followed your suggestions but this need modify the `traced_input` too.